moulinette/authenticators/__init__.py

```diff
@@ -135,33 +144,46 @@ def _open_sessionfile(self, session_id, mode='r'):
         return open_cachefile('%s.asc' % session_id, mode,
                               subdir='session/%s' % self.name)
 
-    def _store_session(self, session_id, session_hash, password):
-        """Store a session and its associated password"""
-        gpg = gnupg.GPG()
-        gpg.encoding = 'utf-8'
-
-        # Encrypt the password using the session hash
-        s = str(gpg.encrypt(password, None, symmetric=True, passphrase=session_hash))
-        assert len(s), "For some reason GPG can't perform encryption, maybe check /root/.gnupg/gpg.conf or re-run with gpg = gnupg.GPG(verbose=True) ?"
+    def _store_session(self, session_id, session_token):
+        """Store a session to be able to use it later to reauthenticate"""
 
+        # We store a hash of the session_id and the session_token (the token is assumed to be secret)
+        to_hash = "{id}:{token}".format(id=session_id, token=session_token)
+        hash_ = hashlib.sha256(to_hash).hexdigest()
         with self._open_sessionfile(session_id, 'w') as f:
-            f.write(s)
+            f.write(hash_)
 
-    def _retrieve_session(self, session_id, session_hash):
+    def _authenticate_session(self, session_id, session_token):
         """Retrieve a session and return its associated password"""
         try:
+            # FIXME : shouldn't we also add a check that this session file
+            # is not too old ? e.g. not older than 24 hours ? idk...
+
             with self._open_sessionfile(session_id, 'r') as f:
-                enc_pwd = f.read()
+                stored_hash = f.read()
         except IOError as e:
             logger.debug("unable to retrieve session", exc_info=1)
             raise MoulinetteError('unable_retrieve_session', exception=e)
         else:
-            gpg = gnupg.GPG()
-            gpg.encoding = 'utf-8'
-
-            decrypted = gpg.decrypt(enc_pwd, passphrase=session_hash)
-            if decrypted.ok is not True:
-                error_message = "unable to decrypt password for the session: %s" % decrypted.status
-                logger.error(error_message)
-                raise MoulinetteError('unable_retrieve_session', exception=error_message)
-            return decrypted.data
+            to_hash = "{id}:{token}".format(id=session_id, token=session_token)
+            hash_ = hashlib.sha256(to_hash).hexdigest()
+
+            if hash_ != stored_hash:
+                raise MoulinetteError('invalid_token')
```

Comment:
Note that this is the critical crypto part where we check the user provided id+token against a hashed id+token stored locally during the initial auth.

Reply:
It's not super clear to understand what is supposed to be the token, the hash, etc. :)

Reply:
Eh yup, I added some big comment (dunno if that's the right place for it) to explain what's happening.